R/util_dfc.R

```diff
@@ -0,0 +1,238 @@
+##########################################################################
+###
+###     util_dfc.R
+###     Dan Greene, 12/21/17
+###
+###     The goal of the util_dfc ("data frame compare") module is to make it easy to see
+###     where two data frames differ for debugging and further analysis.
+###
+###     Depends on util.R
+###
+
+util_dfc.compare_vecs <- function(vec1, vec2) {
+  # compare vector elements without regard to order and without checking for duplicates.
+  # return a list with three components -
+  ## shared_elements: the shared elements,
+  ## only_in_first: the elements in vec1 and not vec2,
+  ## only_in_second: the elements in vec2 and not vec1.
+  return(list(shared_elements = intersect(vec1, vec2),
+              only_in_first = setdiff(vec1, vec2),
+              only_in_second = setdiff(vec2, vec1)))
+}
+
+
+util_dfc.get_condensed_ids <- function(df, id_cols) {
+  # Helper function for getting a vector of IDs (not necessarily unique) from a DF.
+  # If there are multiple ID columns, they need to be concatenated with "__".
+  if(length(id_cols) == 1) {
+    return(as.character(df[, id_cols]))
+  } else {
+    return(apply(df[, id_cols], 1, paste, collapse = "__"))
+  }
+}
+
+
+util_dfc.compare_colnames <- function(df1, df2) {
+  # compare the column names of two data frames using util_dfc.compare_vecs.
+  # return a list with three components -
+  ## shared_elements: the shared names,
+  ## only_in_first: the names in df1 and not df2,
+  ## only_in_second: the names in df2 and not df1
+
+  # Sanity checks
+  if(!is.data.frame(df1) | !is.data.frame(df2)) {
+    stop("Error - at least one of the first two arguments for util_dfc.compare_colnames is not a data frame.")
+  }
+
+  return(util_dfc.compare_vecs(names(df1), names(df2)))
+}
+
+
+
+
+util_dfc.compare_identifiers <- function(df1, df2, id_cols, id_cols_uniquely_identify_rows = FALSE) {
+  # Compare two dfs on a set of one or more commonly-shared columns that identify rows.
+  # If rows are identified by multiple columns (e.g. team x student x week), include all column names in id_cols.
+  # Return a list with shared IDs, IDs in df1 only, and IDs in df2 only.
+  #
+  # If you expect id_cols to UNIQUELY identify each row in both dfs, set id_cols_uniquely_identify_rows to TRUE.
+  # Otherwise, duplicate ID values within each data frame will get ignored.
```

Comment:
I'm not sure I understand this comment. Ignored, how? What is the opposite of ignoring? Is the opposite to include duplicates as un-shared identifiers if they are duplicated in one df but not the other?

If so, then it seems to me like this is conceptually different from evaluating whether the identifiers in one df are contained in the other, and it might make sense to separate that into a separate function.

Reply:
in fact, I have a function that checks unique indices that I copy and paste a lot—perhaps all or part could be useful here

```
check_index_is_unique <- function(df, index_cols = NULL){
    # Checks whether the vector of column names index_cols define 
    # unique rows in df by comparing the number of rows in the whole df
    # to the number of rows defined by just the unique combinations of the 
    # index_cols. Returns boolean if throw_error is FALSE, otherwise simply
    # throws an error if the check fails.
    
    # if no index cols are specified, then check whether the whole df is unique
    if(is.null(index_cols)){
        index_cols <- names(df)
    }
    
    # Throw an error if df is missing any index_cols
    if(any(!index_cols %in% names(df))){
        stop("in check_index_is_unique, one of your index cols does not exist in df.")
    }
    
    nrows_index_cols <- df %>%
        select(one_of(index_cols)) %>%
        unique %>%
        nrow
    
    nrows_whole_df <- nrow(df)
    
    return(nrows_index_cols == nrows_whole_df)
}
```

Reply:
Yeah this is very helpful and my function is not clear. I'd like to talk on Monday about how we'd like all of these functions to handle duplicates.

Reply:
looking forward to it! :)